fendermint/vm/snapshot/src/client.rs

```diff
@@ -0,0 +1,151 @@
+    ) -> Stm<Option<SnapshotItem>> {
+        let mut snapshots = self.state.snapshots.read_clone()?;
+        let mut snapshot = None;
+        for s in snapshots.iter_mut() {
+            if s.manifest.block_height == block_height && s.manifest.version == version {
+                s.last_access = SystemTime::now();
+                snapshot = Some(s.clone());
+                break;
+            }
+        }
+        if snapshot.is_some() {
+            self.state.snapshots.write(snapshots)?;
+        }
+        Ok(snapshot)
+    }
+
+    /// If the offered snapshot is accepted, we create a temporary directory to hold the chunks
+    /// and remember it as our current snapshot being downloaded.
+    pub fn offer_snapshot(&self, manifest: SnapshotManifest) -> StmResult<(), SnapshotError> {
+        if manifest.version != 1 {
+            abort(SnapshotError::IncompatibleVersion(manifest.version))
+        } else {
+            match tempdir() {
+                Ok(dir) => {
+                    let download = SnapshotDownload {
+                        manifest,
+                        download_dir: Arc::new(dir),
+                        next_index: TVar::new(0),
+                    };
+                    self.state.current_download.write(Some(download))?;
+                    Ok(())
+                }
+                Err(e) => abort(SnapshotError::from(e))?,
+            }
+        }
+    }
+
+    /// Take a chunk sent to us by a remote peer. This is our chance to validate chunks on the fly.
+    ///
+    /// Return a flag indicating whether all the chunks have been received and loaded to the blockstore.
+    pub fn apply_chunk(&self, index: u32, contents: Vec<u8>) -> StmResult<bool, SnapshotError> {
+        if let Some(cd) = self.state.current_download.read()?.as_ref() {
+            let next_index = cd.next_index.read_clone()?;
+            if index != next_index {
+                abort(SnapshotError::UnexpectedChunk(next_index, index))
+            } else {
+                let part_path = cd
+                    .download_dir
+                    .as_ref()
+                    .path()
+                    .join(format!("{}.part", index));
+
+                // We are doing IO inside the STM transaction, but that's okay because there is no contention on the download.
+                match std::fs::write(part_path, contents) {
+                    Ok(()) => {
+                        let next_index = index + 1;
+                        cd.next_index.write(next_index)?;
+
+                        if next_index == cd.manifest.chunks {
+                            // Verify the checksum then load the snapshot and remove the current download from memory.
+                            match manifest::parts_checksum(cd.download_dir.as_ref()) {
+                                Ok(checksum) => {
+                                    if checksum == cd.manifest.checksum {
+                                        // TODO: Import Snapshot.
```

Comment:
I'm not sure yet if we should import the snapshot here, or make another call from the `App`. In any case, the `App` will have to write to its app state history, so we'll have to return the full `SnapshotManifest` so it can construct the necessary record.

Reply:
This change was done in https://github.com/consensus-shipyard/fendermint/pull/443/commits/6e3c5873d536a858e5b255e7f4198fbbadc87ab2

Reply:
I had a few questions on how you import and if you already persist this snapshot so it is available to offer it to other peers, but I see this is addressed in the follow-up PR. Will ask my questions there if they are still not solved. :pray: